### DISCUSSION TITLE
The stereoKit graphics rendering rate is a bit slow!

### DISCUSSION THREAD

**Discussion by suzhouhu** [StereoKit/StereoKit#381] [Q&A]
![图片](https://user-images.githubusercontent.com/23079693/172056584-87c32f58-b5eb-4f95-a514-4455687ff722.png)

It took **five minutes** to simulate this polymer! However the same webxr only took **0.2 seconds**

**Comment by maluoi** [comment]
I would not expect a scene like this to give StereoKit trouble! SK's renderer is pretty efficient and fast, but there may be some things you're doing in the code that may trip it up.

SK's renderer is an **_instanced_** renderer, which means it is at its best when rendering the _same_ asset many times. In this scene, it looks like you should have 2 `Mesh` assets, and 1 `Material` (maybe 5 `Material`s if you're coloring the `Material` rather than the draw instance), which would result in 2-10 draw-calls.

If you are creating new `Material`s or generating new `Mesh`es every frame, or per visible item, then SK won't be able to batch any of the rendering.

**Reply by suzhouhu** [reply]
I found that the small file went well, but when the PDB file reaches **500KB**, it is a bit powerless. Maybe there is a bug in my code. I have attached the source code to the attachment. Please help me analyze it if you have time. Thank you very  much.[](url)

**Reply by maluoi** [reply]
I'm afraid I don't see any attachments here!

**Comment by suzhouhu** [comment]
I'm sorry, I couldn't add it successfully just now, and now I have added an email attachment



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "StereoKit/StereoKit"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年6月6日(星期一) 凌晨1:15
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [StereoKit/StereoKit] The stereoKit graphics rendering rate is a bit slow! (Discussion #381)





 
I would not expect a scene like this to give StereoKit trouble! SK's renderer is pretty efficient and fast, but there may be some things you're doing in the code that may trip it up.
 
First off, SK's renderer is an instanced renderer, which means it is at its best when rendering the same asset many times. In this scene, it looks like you should have 2 Mesh assets, and 1 Material (maybe 5 Materials if you're coloring the Material rather than the draw instance), which would result in 2-10 draw-calls.
 
If you are creating new Materials or generating new Meshes every frame, or per visible item, then SK won't be able to batch any of the rendering.

**Reply by maluoi** [reply]
An email attachment won't work with GitHub, you'll either need to host it as a link somewhere, or put it up on something like [gist](https://gist.github.com/)

**Comment by suzhouhu** [comment]


**Reply by maluoi** [reply]
Looking at the code for this, you are indeed generating a Mesh and Material asset for every single node.

So from what I wrote above:
>SK's renderer is an instanced renderer, which means it is at its best when rendering the same asset many times. In this scene, it looks like you should have 2 Mesh assets, and 1 Material (maybe 5 Materials if you're coloring the Material rather than the draw instance), which would result in 2-10 draw-calls.
>
>If you are creating new Materials or generating new Meshes every frame, or per visible item, then SK won't be able to batch any of the rendering.

In this code, you'll perhaps have thousands of Meshes and Materials, which is _really_ bad for rendering. Try and add some code for re-using your assets! For example, you should only ever call this code once in the entire application: `Mesh.GenerateCylinder(1, 1, Vec3.Forward);`

**Comment by suzhouhu** [comment]
How to do？



发自我的iPhone


------------------ Original ------------------
From: Nick Klingensmith ***@***.***&gt;
Date: Mon,Jun 6,2022 2:16 PM
To: StereoKit/StereoKit ***@***.***&gt;
Cc: 刘鹄中 ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [StereoKit/StereoKit] The stereoKit graphics rendering rate is a bit slow! (Discussion #381)





 
Looking at the code for this, you are indeed generating a Mesh and Material asset for every single node.
 
So from what I wrote above:
  
SK's renderer is an instanced renderer, which means it is at its best when rendering the same asset many times. In this scene, it looks like you should have 2 Mesh assets, and 1 Material (maybe 5 Materials if you're coloring the Material rather than the draw instance), which would result in 2-10 draw-calls.
 
If you are creating new Materials or generating new Meshes every frame, or per visible item, then SK won't be able to batch any of the rendering.
  
In this code, you'll perhaps have thousands of Meshes and Materials, which is really bad for rendering. Try and add some code for re-using your assets! For example, you should only ever call this code once in the entire application: Mesh.GenerateCylinder(1, 1, Vec3.Forward);

**Reply by maluoi** [reply]
This is basically just like caching expensive computations:

```csharp
Mesh cylinder = Mesh.GenerateCylinder(1, 1, Vec3.Forward);
Mesh sphere = Mesh.GenerateSphere(1);
Material matRed = new Material(Shader.Default);
mat.SetColor("color", new Color(1,0,0,1));

Model mdl = new Model();
foreach (var node1 in AllNodes) {
    mdl.AddNode("NodeName", Matrix.TRS(...), sphere, matRed);
    ...
}
```

**Reply by suzhouhu** [reply]
I followed your method to set the Material and Mesh at **one time.** However, there was basically no change. When I opened the macromolecule file, it still took **three minutes** of precious time! Could you please see if there are other optimization measures? I attach the source code: [https://3dchem.cloud/PDBReader.zip](https://3dchem.cloud/PDBReader.zip)

**Reply by maluoi** [reply]
Okay, I see! So this is not a rendering issue, but an initialization issue. Rendering is fine and fast here!

First thing, I'd recommend you get familiar with Visual Studio's profiling tools. VS has some really excellent tools for checking where performance issues are, so you can get a lot of good information with just a little bit of effort! You can find it under `Debug->Performance Profiler`, I'd definitely recommend playing around with it :)

Here's what I get when running it.
![image](https://user-images.githubusercontent.com/8484884/172291021-e222a0d5-ba6e-4230-9c1e-9a2e3f87c839.png)

It looks like the core issue here is `Model.AddNode`. I'll have to look a little deeper, but there is some code that recalculates the `Model.Bounds` each time a ModelNode is added. It's definitely not ideal for this situation, I can definitely optimize this for the next update!

Ultimately, I think it may be a good idea to look into representing this without using `Model`. `Model` unfortunately will always come with a small performance overhead from Matrix multiplications. Each node in a Model has a Model Space transform, and when you draw the Model, SK has to convert that to World Space, which involves a Matrix multiply.

You can skip that extra cost by storing your data in World space to begin with! The Hierarchy class can help here. Doing it this way also makes it easier to use the instance color instead of having Materials per color, which is nice. Here's some pseudocode!

```csharp
struct MoleculeElement {
    public Mesh mesh;
    public Matrix matrix;
    public Color color;
    public MoleculeElement(Mesh m, Matrix x, Color c) {mesh = m; matrix = x; color = c;}
    public void Draw() => mesh.Draw(Default.Material, matrix, color);
}

List<MoleculeElement> elements = new();

Matrix curr = Matrix.S(0.1);
Hierarchy.Push(curr);
    elements.Add(new(Mesh.Sphere, Hierarchy.ToWorld(), Color.White));

    curr = Matrix.T(1,0,0);
    Hierarchy.Push(curr);
        elements.Add(new(Mesh.Sphere, Hierarchy.ToWorld(), new Color(1,0,0,1));
    Hierarchy.Pop();
Hierarchy.Pop();

elements.ForEach( e=>e.Draw() );
```

**Reply by maluoi** [reply]
I've noted down the performance issue as a task over here: https://github.com/StereoKit/StereoKit/issues/383

I'll probably get to it after MR Dev Days :)

**Reply by suzhouhu** [reply]
This method really works! I calculated that it took about 18 seconds to load this large molecule. A lot of progress! However, my WEBXE: [https://www.3dchem.cn/WEBVR/](https://www.3dchem.cn/WEBVR/) also loaded this molecule, it only took 0.2 seconds, it seems that it needs improvement
